File: material-validator.py
import pandas as pd
import re
import time
from typing import List, Tuple, Optional, Dict

class MaterialValidator:
    """
    A high-speed, deterministic validator for material composition strings.

    This validator operates on the fundamental rule that the compositions of any
    given "part" of a material must sum to 100. It makes no prior assumptions
    about the format, keywords, or delimiters used.
    """
    def __init__(self):
        # A compiled regex for high-speed tokenization.
        # It finds:
        # 1. Floating-point or integer numbers.
        # 2. Sequences of letters (words).
        # 3. Any single character that isn't a letter, number, or whitespace (delimiters/symbols).
        self.tokenizer_regex = re.compile(r'(\d+(?:\.\d+)?|[a-zA-Z]+|[^a-zA-Z0-9\s])')
        self.validation_cache: Dict[str, bool] = {}

    def _tokenize(self, s: str) -> List[str]:
        """Tokenizes a raw string into its fundamental components."""
        return self.tokenizer_regex.findall(s)

    def _validate_part(self, tokens: List[str]) -> bool:
        """
        Checks if a given sequence of tokens represents a valid part
        by summing its numerical components.
        """
        # Quick check: if there are no numbers, it can't be a valid part.
        if not any(re.fullmatch(r'\d+(?:\.\d+)?', t) for t in tokens):
            return False
            
        total = sum(float(t) for t in tokens if re.fullmatch(r'\d+(?:\.\d+)?', t))
        
        # Use a small tolerance for floating-point comparisons.
        return abs(total - 100.0) < 1e-6

    def _find_valid_partition(self, tokens: List[str]) -> Optional[List[List[str]]]:
        """
        Hypothesizes and tests different ways to partition the token list
        to satisfy the 100-sum rule for all sub-parts.
        """
        # Hypothesis 1 (Most Common): The entire string is a single part.
        if self._validate_part(tokens):
            return [tokens]

        # Hypothesis 2: The string is multi-part, split by common delimiters.
        # This is assumption-free as it only succeeds if the resulting parts are valid.
        potential_delimiters = ['//', '|']
        for delim in potential_delimiters:
            if delim in tokens:
                try:
                    parts = []
                    current_part = []
                    for token in tokens:
                        if token == delim:
                            parts.append(current_part)
                            current_part = []
                        else:
                            current_part.append(token)
                    parts.append(current_part)

                    if all(self._validate_part(p) for p in parts):
                        return parts
                except Exception:
                    continue # Parsing failed, this hypothesis is wrong.

        # Hypothesis 3: The string is multi-part, split by "KEYWORD:" patterns.
        # This is a very common structure in product data.
        try:
            parts = []
            current_part = []
            i = 0
            while i < len(tokens):
                # Check for the WORD followed by ':' pattern
                if (i + 1 < len(tokens) and 
                    re.fullmatch(r'[a-zA-Z]+', tokens[i]) and 
                    tokens[i+1] == ':'):
                    
                    if current_part: # If we have a part accumulated
                        parts.append(current_part)
                    current_part = [tokens[i], tokens[i+1]] # Start new part with keyword
                    i += 2
                else:
                    current_part.append(tokens[i])
                    i += 1
            if current_part:
                parts.append(current_part)

            if len(parts) > 1 and all(self._validate_part(p) for p in parts):
                return parts
        except Exception:
            pass

        # If no valid partition could be found
        return None

    def validate(self, material_string: str) -> bool:
        """
        Validates a single material string. Uses a cache for speed.
        """
        if material_string in self.validation_cache:
            return self.validation_cache[material_string]

        tokens = self._tokenize(material_string)
        if not tokens:
            self.validation_cache[material_string] = False
            return False
            
        result = self._find_valid_partition(tokens) is not None
        self.validation_cache[material_string] = result
        return result

    def bulk_validate(self, material_list: List[str]) -> Tuple[List[str], List[str]]:
        """
        Validates a list of material strings and separates them into valid and invalid lists.
        """
        valid_materials = []
        invalid_materials = []
        for material in material_list:
            if self.validate(material):
                valid_materials.append(material)
            else:
                invalid_materials.append(material)
        return valid_materials, invalid_materials


if __name__ == '__main__':
    # --- 1. Load Data ---
    try:
        df = pd.read_csv('esqualo_2022_fall.csv')
        materials_list = df['material'].dropna().unique().tolist()
        print(f"Loaded {len(materials_list)} unique material strings for analysis from 'esqualo_2022_fall.csv'.\n")
    except FileNotFoundError:
        print("Error: 'esqualo_2022_fall.csv' not found.")
        materials_list = []

    if materials_list:
        # --- 2. Initialize and Run Validator ---
        validator = MaterialValidator()
        
        start_time = time.time()
        valid, invalid = validator.bulk_validate(materials_list)
        end_time = time.time()
        
        # --- 3. Report the Results ---
        print(f"Validation completed in {end_time - start_time:.4f} seconds.\n")
        
        print("="*50)
        print("          VALID Material Strings")
        print("="*50)
        if not valid:
            print("No valid materials found.")
        else:
            print(f"Found {len(valid)} valid material formats:\n")
            for item in sorted(valid):
                print(f"  - '{item}'")

        print("\n\n" + "="*50)
        print("         INVALID Material Strings")
        print("="*50)
        if not invalid:
            print("No invalid materials found.")
        else:
            print(f"Found {len(invalid)} materials that failed validation:\n")
            for item in sorted(invalid):
                print(f"  - '{item}'")

File: go-implementation/material-validator.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"log"
	"math"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

// MaterialValidator holds the logic for validating material strings.
type MaterialValidator struct {
	tokenizerRegex *regexp.Regexp
	validationCache map[string]bool
}

// NewMaterialValidator creates and initializes a new validator.
func NewMaterialValidator() *MaterialValidator {
	// A compiled regex for high-speed tokenization.
	regex := regexp.MustCompile(`(\d+(?:\.\d+)?|[a-zA-Z]+|[^a-zA-Z0-9\s])`)
	return &MaterialValidator{
		tokenizerRegex: regex,
		validationCache: make(map[string]bool),
	}
}

// tokenize breaks a raw string into its fundamental components.
func (v *MaterialValidator) tokenize(s string) []string {
	return v.tokenizerRegex.FindAllString(s, -1)
}

// isNumeric checks if a string can be parsed as a number.
func isNumeric(s string) bool {
	_, err := strconv.ParseFloat(s, 64)
	return err == nil
}

// validatePart checks if a given sequence of tokens sums to 100.
func (v *MaterialValidator) validatePart(tokens []string) bool {
	var total float64
	hasNumbers := false
	for _, t := range tokens {
		if isNumeric(t) {
			hasNumbers = true
			num, err := strconv.ParseFloat(t, 64)
			if err == nil {
				total += num
			}
		}
	}
	return hasNumbers && math.Abs(total-100.0) < 1e-6
}

// findValidPartition hypothesizes and tests different partitions of the token list.
func (v *MaterialValidator) findValidPartition(tokens []string) [][]string {
	// Hypothesis 1: The entire string is a single part.
	if v.validatePart(tokens) {
		return [][]string{tokens}
	}

	// Hypothesis 2: Split by common delimiters.
	potentialDelimiters := []string{"//", "|"}
	for _, delim := range potentialDelimiters {
		if contains(tokens, delim) {
			parts := splitTokens(tokens, delim)
			if len(parts) > 1 && allPartsValid(parts, v.validatePart) {
				return parts
			}
		}
	}

	// Hypothesis 3: Split by "KEYWORD:" patterns.
	parts, isMultiPart := splitByKeywordColon(tokens)
	if isMultiPart && len(parts) > 1 && allPartsValid(parts, v.validatePart) {
		return parts
	}

	return nil
}

// Validate checks a single material string, using a cache for speed.
func (v *MaterialValidator) Validate(materialString string) bool {
	if cachedResult, found := v.validationCache[materialString]; found {
		return cachedResult
	}

	tokens := v.tokenize(materialString)
	if len(tokens) == 0 {
		v.validationCache[materialString] = false
		return false
	}

	result := v.findValidPartition(tokens) != nil
	v.validationCache[materialString] = result
	return result
}

// --- Helper functions to make the logic cleaner ---

func contains(slice []string, item string) bool {
	for _, a := range slice {
		if a == item {
			return true
		}
	}
	return false
}

func splitTokens(tokens []string, delim string) [][]string {
	var parts [][]string
	currentPart := []string{}
	for _, token := range tokens {
		if token == delim {
			if len(currentPart) > 0 {
				parts = append(parts, currentPart)
			}
			currentPart = []string{}
		} else {
			currentPart = append(currentPart, token)
		}
	}
	if len(currentPart) > 0 {
		parts = append(parts, currentPart)
	}
	return parts
}

func allPartsValid(parts [][]string, validateFunc func([]string) bool) bool {
	for _, p := range parts {
		if !validateFunc(p) {
			return false
		}
	}
	return true
}

func splitByKeywordColon(tokens []string) (parts [][]string, isMultiPart bool) {
	currentPart := []string{}
	i := 0
	for i < len(tokens) {
		if i+1 < len(tokens) && regexp.MustCompile(`^[a-zA-Z]+$`).MatchString(tokens[i]) && tokens[i+1] == ":" {
			isMultiPart = true
			if len(currentPart) > 0 {
				parts = append(parts, currentPart)
			}
			currentPart = []string{tokens[i], tokens[i+1]}
			i += 2
		} else {
			currentPart = append(currentPart, tokens[i])
			i++
		}
	}
	if len(currentPart) > 0 {
		parts = append(parts, currentPart)
	}
	return parts, isMultiPart
}


func main() {
	// --- 1. Load Data ---
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run main.go <path_to_csv_file>")
		os.Exit(1)
	}
	filePath := os.Args[1]

	file, err := os.Open(filePath)
	if err != nil {
		log.Fatalf("Error: '%s' not found.", filePath)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	header, err := reader.Read()
	if err != nil {
		log.Fatalf("Error reading header from '%s': %v", filePath, err)
	}

	materialColIndex := -1
	for i, colName := range header {
		if strings.ToLower(colName) == "material" {
			materialColIndex = i
			break
		}
	}

	if materialColIndex == -1 {
		log.Fatal("Error: 'material' column not found.")
	}

	uniqueMaterials := make(map[string]struct{})
	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			continue // Skip rows with parsing errors
		}
		if len(record) > materialColIndex && record[materialColIndex] != "" {
			uniqueMaterials[record[materialColIndex]] = struct{}{}
		}
	}
	
	materialsList := make([]string, 0, len(uniqueMaterials))
	for material := range uniqueMaterials {
		materialsList = append(materialsList, material)
	}
	fmt.Printf("Loaded %d unique material strings for analysis from '%s'.\n\n", len(materialsList), filePath)
	
	// --- 2. Initialize and Run Validator ---
	validator := NewMaterialValidator()
	startTime := time.Now()

	var valid, invalid []string
	for _, material := range materialsList {
		if validator.Validate(material) {
			valid = append(valid, material)
		} else {
			invalid = append(invalid, material)
		}
	}
	
	duration := time.Since(startTime)

	// --- 3. Report the Results ---
	fmt.Printf("Validation completed in %v seconds.\n\n", duration.Seconds())
	
	sort.Strings(valid)
	sort.Strings(invalid)
	
	fmt.Println(strings.Repeat("=", 50))
	fmt.Println("          VALID Material Strings")
	fmt.Println(strings.Repeat("=", 50))
	if len(valid) == 0 {
		fmt.Println("No valid materials found.")
	} else {
		fmt.Printf("Found %d valid material formats:\n\n", len(valid))
		for _, item := range valid {
			fmt.Printf("  - '%s'\n", item)
		}
	}

	fmt.Println("\n\n" + strings.Repeat("=", 50))
	fmt.Println("         INVALID Material Strings")
	fmt.Println(strings.Repeat("=", 50))
	if len(invalid) == 0 {
		fmt.Println("No invalid materials found.")
	} else {
		fmt.Printf("Found %d materials that failed validation:\n\n", len(invalid))
		for _, item := range invalid {
			fmt.Printf("  - '%s'\n", item)
		}
	}
}
